src/main/java/org/broadinstitute/hellbender/engine/PathSpecifier.java

```diff
@@ -0,0 +1,260 @@
+            return false;
+        }
+    }
+
+    /**
+     * Resolve the URI to a {@link Path} object.
+     *
+     * @return the resulting {@code Path}
+     * @throws UserException if an I/O error occurs when creating the file system
+     */
+    @Override
+    public Path toPath() {
+        if (getCachedPath() != null) {
+            return getCachedPath();
+        } else {
+            final Path tmpPath = Paths.get(getURI());
+            setCachedPath(tmpPath);
+            return tmpPath;
+        }
+    }
+
+    @Override
+    public String getToPathFailureReason() {
+        if (pathFailureReason == null) {
+            try {
+                toPath();
+                return String.format("'%s' appears to be a valid Path", rawInputString);
+            } catch (ProviderNotFoundException e) {
+                return String.format("ProviderNotFoundException: %s", e.getMessage());
+            } catch (FileSystemNotFoundException e) {
+                return String.format("FileSystemNotFoundException: %s", e.getMessage());
+            } catch (IllegalArgumentException e) {
+                return String.format("IllegalArgumentException: %s", e.getMessage());
+            } catch (UserException e) {
+                return String.format("UserException: %s", e.getMessage());
+            }
+        }
+        return pathFailureReason;
+    }
+
+    @Override
+    public InputStream getInputStream() {
+        if (!isPath()) {
+            throw new UserException(getToPathFailureReason());
+        }
+
+        final Path resourcePath = toPath();
+        try {
+            return Files.newInputStream(resourcePath);
+        } catch (IOException e) {
+            throw new UserException(
+                    String.format("Could not create open input stream for %s (as URI %s)", getRawInputString(), getURIString()), e);
+        }
+    }
+
+    @Override
+    public OutputStream getOutputStream() {
+        if (!isPath()) {
+            throw new UserException(getToPathFailureReason());
+        }
+
+        final Path resourcePath = toPath();
+        try {
+            return Files.newOutputStream(resourcePath);
+        } catch (IOException e) {
+            throw new UserException(String.format("Could not open output stream for %s (as URI %s)", getRawInputString(), getURIString()), e);
+        }
+    }
+
+    // get the cached path associated with this URI if its already been created
+    protected Path getCachedPath() { return cachedPath; }
+
+    protected void setCachedPath(Path path) {
+        this.cachedPath = path;
+    }
+
+    @Override
+    public String toString() {
+        return rawInputString;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (!(o instanceof PathSpecifier)) return false;
+
+        PathSpecifier that = (PathSpecifier) o;
+
+        if (!getURIString().equals(that.getURIString())) return false;
```

Comment:
I think I agree. I have mixed feelings since different input strings can result in the same URI. But I think you're right that the stricter equality is probably better.

Reply:
I meant more that it was repetitive to compare the uriString AND the URI itself since those should hopefully be equivalent?  It might make sense to have equality be on the normalized URI and ignore the raw String.  It might also make sense not to define equality if we can't decide what it means unambiguously.

Reply:
Right - it is repetitive.`FeatureManager` for example uses `FeatureInputs` as `Map` keys,  so I think we should put some stake in the ground about equality. My original implementation would have treated different input strings that resolve to the same URI as the equal, but I think the stricter implementation (treat them as different) makes more sense. If some caller wants to use URI equality, it do so with the use the URIs ?